### DISCUSSION TITLE
EGPv4 Concept Thread

### DISCUSSION THREAD

**Discussion by Denneisk** [wiremod/wire#2849] [General]
Using this thread to write down ideas for a potential EGPv4. With E2 changing so much, EGP should have a similar update to bring it to a new age. Suggestions are greatly appreciated.

My main idea for EGPv4 is to make it more like a 2D engine that runs on E2.

- Basic physics engine (collision, velocity, friction, etc)
- Built-in user-interactive objects such as buttons, text entries
- Responsive design, relative sizing
- Clientside E2/EGP script
- Textures and sprites, with an editor
- Panels-like system, think more like frames but reactive
- Objects can have draw/tick callbacks added to them
- Markup language elements

**Comment by bjcscat** [comment]
I think the overall model could include a single EGP state observed by potentially multiple EGP screens.
This state could be transmitted to the client for purposes of smoothing animations or other gradual movements.
Differences in screen size or other information could be handled with a new data type that include scaled and explicit size parts.

**Comment by vurvdev** [comment]
The hml (hud markup language) part of wire extras is pretty interesting, maybe some inspiration can be taken from it

https://github.com/wiremod/wire-extras/issues/114

(Note it's probably completely broken)

**Comment by ReallyWeirdCat** [comment]
Maybe some sort of built-in animations for objects, like simple fade-in and fade-out. Or importing pngs from the web, that would be pretty good.

**Comment by deltamolfar** [comment]
- EGP Events. (Examples below):
`event egpHover(EGP, Index/Object, Player, IsHovered)` -> Called when any player hovers on egp object. Maybe for optimization purposes - you should enable hover callbacks in object's properties.
`event egpClicked(EGP, Index/Object, Player)` -> Called when any player clicks on egp object. Maybe for optimization purposes - you should enable hover callbacks in object's properties.
- EGP Object clipping -> Making possible to "cut" one shape from another, or just implement GLua's stencil-like functions

**Reply by lennyb365** [reply]
Love these suggestions
